File: CC-Mastery/sdk_demo/layer3/01_query_basics.ts
/**
 * Layer 3, Level 1: query() Basics in TypeScript
 * =================================================
 * Run OUTSIDE Claude Code session:
 *     cd sdk_demo/layer3 && npx tsx 01_query_basics.ts
 *
 * KEY DIFFERENCES FROM PYTHON:
 *   Python: query() → AsyncGenerator[Message]
 *   TS:     query() → Query (AsyncGenerator<SDKMessage> + control methods)
 *
 *   Python messages: AssistantMessage, ResultMessage, SystemMessage, UserMessage
 *   TS messages: SDKAssistantMessage, SDKResultMessage, SDKSystemMessage, SDKUserMessage
 *
 *   Python content: TextBlock, ToolUseBlock
 *   TS content: msg.message.content[] (Anthropic BetaMessage format)
 */

import { query } from "@anthropic-ai/claude-agent-sdk";
import type {
  SDKMessage,
  SDKSystemMessage,
  SDKAssistantMessage,
  SDKUserMessage,
  SDKResultMessage,
  SDKResultSuccess,
  SDKResultError,
  Options,
} from "@anthropic-ai/claude-agent-sdk";
import path from "path";

const TESTBED = path.resolve(import.meta.dirname!, "..", "testbed");

async function main() {
  console.log("=".repeat(60));
  console.log("  TypeScript SDK: query() Basics");
  console.log("=".repeat(60));
  console.log(`Target: ${TESTBED}\n`);

  // ---------------------------------------------------------------
  // query() returns a Query object:
  //   - Implements AsyncGenerator<SDKMessage> (for await...of)
  //   - Has control methods: interrupt(), setModel(), etc.
  //
  // Options uses camelCase (not snake_case like Python):
  //   Python: allowed_tools, max_turns, setting_sources
  //   TS:     allowedTools,  maxTurns,  settingSources
  // ---------------------------------------------------------------
  const options: Options = {
    // Tools: read-only analysis
    tools: ["Read", "Glob"],

    // Budget caps
    maxTurns: 5,
    maxBudgetUsd: 0.10,

    // Model
    model: "sonnet",

    // Working directory
    cwd: TESTBED,

    // Isolated: no CLAUDE.md, no hooks
    settingSources: [],

    // Don't save to disk (ephemeral)
    persistSession: false,  // Python: --no-session-persistence
  };

  const msgCounts = { system: 0, assistant: 0, user: 0, result: 0 };

  // ---------------------------------------------------------------
  // Iterate the message stream
  // ---------------------------------------------------------------
  const q = query({
    prompt: `Read all Python files in ${TESTBED} and list the functions in each.`,
    options,
  });

  for await (const message of q) {
    // --- Type narrowing by message.type ---

    if (message.type === "system" && "subtype" in message) {
      const sys = message as SDKSystemMessage;
      if (sys.subtype === "init") {
        msgCounts.system++;
        console.log("─".repeat(60));
        console.log(`SYSTEM (init)`);
        console.log(`  session: ${sys.session_id}`);
        console.log(`  model:   ${sys.model}`);
        console.log(`  tools:   ${sys.tools.join(", ")}`);
      }
    }

    else if (message.type === "assistant") {
      const asst = message as SDKAssistantMessage;
      msgCounts.assistant++;
      console.log("─".repeat(60));
      console.log(`ASSISTANT (model=${asst.message.model})`);

      // Content blocks are Anthropic API format (BetaMessage.content)
      for (const block of asst.message.content) {
        if (block.type === "text") {
          const preview = block.text.length > 200
            ? block.text.slice(0, 200) + "..."
            : block.text;
          console.log(`  TEXT: ${preview}`);
        }
        else if (block.type === "thinking") {
          console.log(`  THINKING: ${block.thinking.slice(0, 100)}...`);
        }
        else if (block.type === "tool_use") {
          console.log(`  TOOL_USE: ${block.name}`);
          console.log(`    id: ${block.id}`);
          console.log(`    input: ${JSON.stringify(block.input).slice(0, 120)}`);
        }
        else if (block.type === "tool_result") {
          console.log(`  TOOL_RESULT: ${JSON.stringify(block).slice(0, 100)}`);
        }
      }
    }

    else if (message.type === "user") {
      msgCounts.user++;
      // UserMessage carries tool results back to the agent
      // Usually not interesting to display
    }

    else if (message.type === "result") {
      const result = message as SDKResultMessage;
      msgCounts.result++;
      console.log("─".repeat(60));
      console.log(`RESULT`);
      console.log(`  subtype:  ${result.subtype}`);
      console.log(`  is_error: ${result.is_error}`);
      console.log(`  turns:    ${result.num_turns}`);
      console.log(`  cost:     $${result.total_cost_usd.toFixed(4)}`);
      console.log(`  duration: ${result.duration_ms}ms`);

      if (result.subtype === "success") {
        const success = result as SDKResultSuccess;
        console.log(`  result:   ${success.result.slice(0, 300)}`);
      } else {
        const error = result as SDKResultError;
        console.log(`  errors:   ${error.errors.join(", ")}`);
      }
    }
  }

  // --- Summary ---
  console.log(`\n${"=".repeat(60)}`);
  console.log(`Message counts:`, msgCounts);
  console.log(`\nKEY DIFFERENCES FROM PYTHON:`);
  console.log(`  Python: isinstance(msg, AssistantMessage)`);
  console.log(`  TS:     msg.type === "assistant"`);
  console.log(`  Python: block.text (TextBlock attribute)`);
  console.log(`  TS:     block.text (object property, after type === "text" check)`);
  console.log(`  Python: ResultMessage.result`);
  console.log(`  TS:     (msg as SDKResultSuccess).result`);
}

main().catch(console.error);

File: CC-Mastery/sdk_demo/layer2/01_query_basics.py
"""
Layer 2, Level 1: query() Basics + Message Flow
=================================================
Run OUTSIDE Claude Code session:
    python sdk_demo/layer2/01_query_basics.py

KEY CONCEPT:
    query() returns an AsyncGenerator that yields typed messages.
    You iterate over them to see EVERYTHING the agent does.

MESSAGE FLOW:
    SystemMessage (init)  →  session_id, tools, model
    AssistantMessage      →  agent thinking + tool calls
    UserMessage           →  tool results fed back to agent
    AssistantMessage      →  agent processes results, may call more tools
    UserMessage           →  more tool results...
    AssistantMessage      →  final text response
    ResultMessage         →  status, cost, structured_output
"""

import asyncio
import sys

sys.stdout.reconfigure(encoding="utf-8")  # Windows fix

from claude_agent_sdk import (
    query,
    ClaudeAgentOptions,
    # --- 4 message types ---
    SystemMessage,    # Init: session_id, tools
    AssistantMessage, # Agent output: text, tool calls
    UserMessage,      # Tool results (auto-generated)
    ResultMessage,    # Final: status, cost, result
    # --- Content block types ---
    TextBlock,        # Agent's text output
    ThinkingBlock,    # Extended thinking (if enabled)
    ToolUseBlock,     # Agent calling a tool: name + input
    ToolResultBlock,  # Tool's response
)
from pathlib import Path

TESTBED = Path(__file__).resolve().parent.parent / "testbed"


async def main():
    print("=" * 60)
    print("  query() Message Flow Demo")
    print("=" * 60)
    print()

    msg_count = {"system": 0, "assistant": 0, "user": 0, "result": 0}

    # ---------------------------------------------------------------
    # query() signature:
    #   query(prompt: str, options: ClaudeAgentOptions) -> AsyncGenerator[Message]
    #
    # The generator yields messages as the agent works.
    # You MUST iterate it — otherwise the agent never runs.
    # ---------------------------------------------------------------
    async for message in query(
        prompt=f"Read {TESTBED / 'utils.py'} and list all function names.",
        options=ClaudeAgentOptions(
            # Minimum tools: read-only
            tools=["Read"],
            # Budget cap
            max_turns=3,
            max_budget_usd=0.10,
            # Model
            model="sonnet",
            # Isolated: no CLAUDE.md, no hooks
            setting_sources=[],
        ),
    ):
        # --- Dispatch on message type ---

        if isinstance(message, SystemMessage):
            msg_count["system"] += 1
            print(f"{'─'*60}")
            print(f"SYSTEM (subtype={message.subtype})")
            # data contains session_id, tools, model, cwd
            sid = message.data.get("session_id", "?")
            print(f"  session_id: {sid}")
            print(f"  data keys: {list(message.data.keys())}")

        elif isinstance(message, AssistantMessage):
            msg_count["assistant"] += 1
            print(f"{'─'*60}")
            print(f"ASSISTANT (model={message.model})")

            for block in message.content:
                if isinstance(block, TextBlock):
                    # The agent's text — reasoning or final answer
                    preview = block.text[:200]
                    if len(block.text) > 200:
                        preview += "..."
                    print(f"  TEXT: {preview}")

                elif isinstance(block, ThinkingBlock):
                    # Extended thinking (only with thinking enabled)
                    print(f"  THINKING: {block.thinking[:100]}...")

                elif isinstance(block, ToolUseBlock):
                    # Agent is calling a tool
                    print(f"  TOOL_USE: {block.name}")
                    print(f"    id: {block.id}")
                    print(f"    input: {_fmt(block.input)}")

                elif isinstance(block, ToolResultBlock):
                    # Tool result embedded in assistant message
                    content = str(block.content)[:100] if block.content else "(empty)"
                    print(f"  TOOL_RESULT: {content}")

        elif isinstance(message, UserMessage):
            msg_count["user"] += 1
            print(f"{'─'*60}")
            print(f"USER (auto-generated tool result)")
            # UserMessage carries tool results back to the agent.
            # You don't create these — the SDK does automatically.
            if message.tool_use_result:
                content = str(message.tool_use_result)[:150]
                print(f"  tool_result: {content}...")
            elif isinstance(message.content, list):
                for block in message.content:
                    if isinstance(block, ToolResultBlock):
                        snippet = str(block.content)[:150] if block.content else ""
                        print(f"  tool_result [{block.tool_use_id[:8]}]: {snippet}...")

        elif isinstance(message, ResultMessage):
            msg_count["result"] += 1
            print(f"{'─'*60}")
            print(f"RESULT")
            print(f"  subtype:   {message.subtype}")
            print(f"  is_error:  {message.is_error}")
            print(f"  turns:     {message.num_turns}")
            print(f"  cost:      ${message.total_cost_usd or 0:.4f}")
            print(f"  duration:  {message.duration_ms}ms")
            print(f"  session:   {message.session_id}")
            if message.result:
                print(f"  result:    {message.result[:300]}")

    # --- Summary ---
    print(f"\n{'='*60}")
    print(f"Message counts: {msg_count}")
    print(f"  Typical flow: 1 system + N*(assistant+user) + 1 result")


def _fmt(d: dict, max_len: int = 120) -> str:
    """Format dict for display, truncating long values."""
    parts = []
    for k, v in d.items():
        vs = str(v)
        if len(vs) > 60:
            vs = vs[:57] + "..."
        parts.append(f"{k}={vs}")
    out = ", ".join(parts)
    return out[:max_len]


if __name__ == "__main__":
    asyncio.run(main())
